js/apps/admin-ui/src/i18n/OverridesBackend.ts

```diff
@@ -0,0 +1,160 @@
+  DEFAULT_NAMESPACE,
+  KEY_SEPARATOR,
+  NAMESPACE_SEPARATOR,
+} from "./i18n";
+
+type ParsedOverrides = { [namespace: string]: { [key: string]: string } };
+
+/** A custom backend that merges the overrides the static labels with those defined by the user in the console. */
+export class OverridesBackend extends HttpBackend {
+  #overridesCache = new Map<string, Promise<ParsedOverrides>>();
+
+  async loadUrl(
+    url: string,
+    callback: ReadCallback,
+    languages?: string | string[],
+    namespaces?: string | string[]
+  ) {
+    try {
+      const [data, overrides] = await Promise.all([
+        this.#loadUrlPromisified(url, languages, namespaces),
+        this.#loadOverrides(languages),
+      ]);
+
+      const namespace = this.#determineNamespace(namespaces);
+      const merged = this.#applyOverrides(namespace, data, overrides);
+
+      callback(null, merged);
+    } catch (error) {
+      callback(error as CallbackError, null);
+    }
+  }
+
+  #applyOverrides(
+    namespace: string,
+    data: ResourceKey,
+    overrides: ParsedOverrides
+  ) {
+    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+    if (typeof data === "string" || !overrides[namespace]) {
+      return data;
+    }
+
+    // Ensure we are operating on a cloned data structure to prevent in place mutations.
+    const target = structuredClone(data);
+
+    for (const [path, value] of Object.entries(overrides[namespace])) {
+      this.#applyOverride(target, path, value);
+    }
+
+    return target;
+  }
+
+  /** Applies an override by converting path segments denoted with a key separator as nested objects and merging the result. */
+  #applyOverride(target: Record<string, any>, path: string, value: string) {
+    const trail = path.split(KEY_SEPARATOR);
+    let pointer = target;
+
+    trail.forEach((segment, index) => {
+      const isLast = index === trail.length - 1;
+      pointer = pointer[segment] = isLast ? value : pointer[segment] ?? {};
+    });
+  }
+
+  #loadOverrides(languages?: string | string[]) {
+    const locale = this.#determineLocale(languages);
+    const cachedOverrides = this.#overridesCache.get(locale);
+
+    if (cachedOverrides) {
+      return cachedOverrides;
+    }
+
+    const overrides = adminClient.realms
+      .getRealmLocalizationTexts({
+        realm: adminClient.realmName,
+        selectedLocale: locale,
+      })
+      .then((data) => this.#parseOverrides(data));
+
+    this.#overridesCache.set(locale, overrides);
+
+    // Evict cached request on failure.
+    overrides.catch((error) => {
+      this.#overridesCache.delete(locale);
+      return Promise.reject(error);
+    });
+
+    return overrides;
+  }
+
+  #parseOverrides(data: Record<string, string>) {
+    const parsed: ParsedOverrides = {};
+
+    for (const [path, value] of Object.entries(data)) {
+      const parts = path.split(NAMESPACE_SEPARATOR);
+      const namespace = parts.length > 1 ? parts[0] : DEFAULT_NAMESPACE;
```

Comment:
Maybe it's less errorprone, when messages without a namespace are ignored.

Then, the correct realm localization key for translation would always be the namespace filename (without extension), followed by the namespace separator and the actual key.

WDYT?

Reply:
Yes, I do actually agree this would make things less complicated. I tried to mimic the conventions of i18next here as much as possible, but perhaps simpler is better. I'll change this implementation.